File: typescript/packages/core/src/workspace/executor/builtins/exec/exec.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult, materialize } from '../../../../io/types.ts'
import type { ByteSource } from '../../../../io/types.ts'
import type { DispatchFn } from '../../../../runtime/types.ts'
import { FD_BOTH, FD_CLOSE, FD_STDERR, FD_STDIN } from '../../../../shell/constants.ts'
import { badDescriptorLine, unsupportedDescriptor } from '../../../../shell/descriptors.ts'
import { type Redirect, RedirectKind } from '../../../../shell/types.ts'
import { fsStrerror, isFsError } from '../../../../utils/errors.ts'
import type { PathSpec } from '../../../../types.ts'
import type { Session } from '../../../session/session.ts'
import { ExecutionNode } from '../../../types.ts'
import { createFile } from '../../create.ts'
import { toScope } from '../scope.ts'
import type { EXEC_STREAM_FIELDS } from './constants.ts'
import { CLOSED, TO_STDERR, TO_STDIN, TO_STDOUT } from './constants.ts'
import type { BuiltinCall, Result } from '../types.ts'

/** The `exec` builtin without redirects: bare `exec` is a no-op that
 * succeeds; `exec CMD` has no OS-process referent and is refused. */
export function handleExecCommand(args: string[], _session: Session): Result {
  if (args.length === 0)
    return [null, new IOResult(), new ExecutionNode({ command: 'exec', exitCode: 0 })]
  const err = new TextEncoder().encode(
    `mirage: exec: ${args[0] ?? ''}: process replacement is not supported ` +
      '(no OS process to replace)\n',
  )
  return [
    null,
    new IOResult({ exitCode: 2, stderr: err }),
    new ExecutionNode({ command: 'exec', exitCode: 2, stderr: err }),
  ]
}

/** bash's line for a redirect target it could not open. */
function errorLine(label: string, err: unknown): Uint8Array {
  const strerror = isFsError(err) ? (fsStrerror(err) ?? '') : ''
  return new TextEncoder().encode(strerror !== '' ? `${label}: ${strerror}\n` : `${label}\n`)
}

/** The shell-attributed refusal of an `exec` redirect line; the line is
 * null once it was written where the line's own stderr redirect pointed,
 * and `out` carries it when that redirect pointed at the terminal's
 * stdout. */
function execFailure(line: Uint8Array | null, out: Uint8Array | null = null): Result {
  return [
    out,
    new IOResult({ exitCode: 1, stderr: line }),
    new ExecutionNode({ command: 'exec', exitCode: 1, ...(line !== null ? { stderr: line } : {}) }),
  ]
}

/**
 * What a descriptor points at right now, named so a dup can copy it: a
 * path with its append flag, `CLOSED`, or one of the terminal's own
 * streams (`&0`, `&1`, `&2`). The terminal streams are named rather than
 * left as null because a dup copies the *target*, not the role: after
 * `exec 1>&2`, fd 1 is the terminal's stderr whatever fd 2 is later
 * pointed at, and `exec 2>&1` after that puts stderr back on the
 * terminal's stderr, as bash does. Stdin is always the read end, so a
 * stream bound to it (`exec 1>&0`) has nowhere to write.
 */
function identity(session: Session, fd: number): [string, boolean] {
  if (fd === FD_STDIN) return [TO_STDIN, false]
  if (fd === FD_STDERR) return [session.execStderr ?? TO_STDERR, session.execStderrAppend]
  return [session.execStdout ?? TO_STDOUT, session.execStdoutAppend]
}

/** Point a writing stream at an identity. A stream on its own terminal
 * end is stored as null, the undiverted state every reader of
 * `execStdout`/`execStderr` already knows. */
function bind(session: Session, fd: number, id: string, append: boolean): void {
  if (fd === FD_STDERR) {
    session.execStderr = id === TO_STDERR ? null : id
    session.execStderrAppend = append
  } else {
    session.execStdout = id === TO_STDOUT ? null : id
    session.execStdoutAppend = append
  }
}

/**
 * Deliver one stream's bytes where its binding points: to the terminal's
 * stdout, to the terminal's stderr, into a file, or nowhere. Returns the
 * bytes for each terminal stream and whether the write failed: a stream
 * bound to stdin (`exec 1>&0`) cannot be written, which is bash's `write
 * error: Bad file descriptor`.
 */
async function route(
  dispatch: DispatchFn,
  session: Session,
  binding: string | null,
  data: Uint8Array,
  own: string,
): Promise<[Uint8Array | null, Uint8Array | null, boolean]> {
  const target = binding ?? own
  if (target === TO_STDOUT) return [data, null, false]
  if (target === TO_STDERR) return [null, data, false]
  if (target === TO_STDIN) return [null, null, true]
  if (target !== CLOSED) await appendTo(dispatch, session, target, data)
  return [null, null, false]
}

type StreamBindings = Pick<Session, (typeof EXEC_STREAM_FIELDS)[number]>

function bindingsOf(session: Session): StreamBindings {
  return {
    execStdout: session.execStdout,
    execStdoutAppend: session.execStdoutAppend,
    execStderr: session.execStderr,
    execStderrAppend: session.execStderrAppend,
    execStdin: session.execStdin,
  }
}

/**
 * Undo a redirect list that failed part-way, the way bash does: it keeps
 * the side effect of opening each earlier target (`exec >f </missing`
 * leaves an empty `f`) but puts every descriptor back where it stood
 * before the line, so an `echo` after it still reaches the terminal. The
 * diagnostic itself goes through the descriptors as they stood at the
 * failure, which is why `exec 2>e </missing` writes it into `e` and
 * `exec 2>&1 </missing` prints it on stdout.
 */
async function rollBack(
  dispatch: DispatchFn,
  session: Session,
  saved: StreamBindings,
  err: Uint8Array,
): Promise<Result> {
  const partial = session.execStderr
  Object.assign(session, saved)
  const [out, errBytes] = await route(dispatch, session, partial, err, TO_STDERR)
  return execFailure(errBytes, out)
}

function scopeOf(target: unknown): PathSpec {
  return typeof target === 'string' ? toScope(target) : (target as PathSpec)
}

/**
 * Point the shell's own streams at files for the rest of the shell. `exec
 * > file` diverts later stdout, `2> file` stderr, `< file` stdin, `>>`
 * appends; `2>&1`/`>&2` copy one target onto the other; `>&-` closes.
 * The output file is opened now, as bash opens it at exec time. A
 * descriptor above 2 (`exec 3>f`, `exec 3>&-`) is refused with `3: Bad
 * file descriptor`: the shell has no descriptor table, and the old
 * fall-through aliased it onto stdout, which is what `exec 3>&-` closed.
 */
export async function installExecRedirects(
  dispatch: DispatchFn,
  session: Session,
  redirects: Redirect[],
): Promise<Result> {
  const badFd = unsupportedDescriptor(redirects)
  if (badFd !== null) return execFailure(badDescriptorLine(badFd))
  const saved = bindingsOf(session)
  const err = await install(dispatch, session, redirects)
  if (err === null)
    return [null, new IOResult(), new ExecutionNode({ command: 'exec', exitCode: 0 })]
  return rollBack(dispatch, session, saved, err)
}

/** Bind the redirects onto the session's streams, in line order. Returns
 * the diagnostic of the first redirect that fails, with every earlier one
 * still bound, which is the state bash reports from. */
async function install(
  dispatch: DispatchFn,
  session: Session,
  redirects: Redirect[],
): Promise<Uint8Array | null> {
  for (const r of redirects) {
    if (typeof r.target === 'number') {
      // Keyed on the descriptor claimed, not the operator's direction:
      // `2<&-` closes stderr and `0>&-` stdin, as in bash. A dup of a
      // descriptor onto itself changes nothing, so `0<&0` keeps the file
      // an earlier `exec <f` bound; another descriptor onto stdin
      // restores the ambient input.
      if (r.fd === FD_STDIN) {
        if (r.target === FD_CLOSE) session.execStdin = new Uint8Array()
        else if (r.target !== FD_STDIN) session.execStdin = null
      } else if (r.target === FD_CLOSE) {
        bind(session, r.fd, CLOSED, false)
      } else {
        bind(session, r.fd, ...identity(session, r.target))
      }
      continue
    }
    if ((r.kind === RedirectKind.STDIN) !== (r.fd === FD_STDIN)) {
      return badDescriptorLine(r.fd)
    }
    const scope = scopeOf(r.target)
    if (r.kind === RedirectKind.STDIN) {
      try {
        const [data] = await dispatch('read', scope)
        session.execStdin = await materialize(data as ByteSource)
      } catch (err) {
        if (!isFsError(err)) throw err
        return errorLine(scope.rawPath, err)
      }
      continue
    }
    const path = scope.virtual
    try {
      if (await openTarget(dispatch, session, scope, r.append)) session.execOpened.add(path)
    } catch (err) {
      if (!isFsError(err)) throw err
      return errorLine(scope.rawPath, err)
    }
    const streams =
      r.fd === FD_BOTH ? ['stdout', 'stderr'] : r.fd === FD_STDERR ? ['stderr'] : ['stdout']
    for (const stream of streams) {
      if (stream === 'stderr') {
        session.execStderr = path
        session.execStderrAppend = r.append
      } else {
        session.execStdout = path
        session.execStdoutAppend = r.append
      }
    }
  }
  return null
}

/**
 * Open an `exec` redirect target, the way bash does at `exec` time:
 * truncating creates the file empty, appending creates it only when it
 * is not already there so an existing one keeps its bytes. Either way
 * the file exists before the next statement runs, which is what makes
 * `exec >> new; test -e new` succeed with nothing written. Returns
 * whether it was written, which is what marks the target opened.
 */
async function openTarget(
  dispatch: DispatchFn,
  session: Session,
  scope: PathSpec,
  append: boolean,
): Promise<boolean> {
  if (append) {
    try {
      await dispatch('stat', scope)
      return false
    } catch (err) {
      if (!isFsError(err)) throw err
    }
  }
  await createFile(dispatch, session, scope, new Uint8Array(0))
  return true
}

/**
 * Send one statement's output where the shell's `exec` bindings point. A
 * stream bound to a file is appended to it (the first write to each
 * target having truncated it at `exec` time), one bound to the other
 * terminal stream crosses over (`exec 2>&1` puts stderr on stdout), a
 * closed one is dropped, and one bound to stdin fails with bash's `write
 * error: Bad file descriptor`, which is reported on stderr through
 * stderr's own binding and makes the statement's status 1. `command` is
 * the statement's recorded line; its first word names the writer in a
 * write error. Returns the stdout that should still bubble up (null once
 * nothing is left for the terminal).
 */
export async function divertStatement(
  dispatch: DispatchFn,
  session: Session,
  stdout: Uint8Array | null,
  io: IOResult,
  command: string,
): Promise<Uint8Array | null> {
  const outParts: Uint8Array[] = []
  const errParts: Uint8Array[] = []
  let failed = false
  if (stdout !== null && stdout.byteLength > 0) {
    const [out, err, unwritable] = await route(
      dispatch,
      session,
      session.execStdout,
      stdout,
      TO_STDOUT,
    )
    if (out !== null) outParts.push(out)
    if (err !== null) errParts.push(err)
    failed = unwritable
  }
  let stderr = await materialize(io.stderr)
  if (failed) {
    const first = command.trim().split(/\s+/)[0]
    const name = first === undefined || first === '' ? 'bash' : first
    stderr = joinBytes([
      stderr,
      new TextEncoder().encode(`${name}: write error: Bad file descriptor\n`),
    ])
    io.exitCode = 1
  }
  if (stderr.byteLength > 0) {
    const [out, err, unwritable] = await route(
      dispatch,
      session,
      session.execStderr,
      stderr,
      TO_STDERR,
    )
    if (out !== null) outParts.push(out)
    if (err !== null) errParts.push(err)
    if (unwritable) io.exitCode = 1
  }
  io.stderr = errParts.length > 0 ? joinBytes(errParts) : null
  return outParts.length > 0 ? joinBytes(outParts) : null
}

function joinBytes(parts: Uint8Array[]): Uint8Array {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.byteLength, 0))
  let offset = 0
  for (const p of parts) {
    out.set(p, offset)
    offset += p.byteLength
  }
  return out
}

async function appendTo(
  dispatch: DispatchFn,
  session: Session,
  target: string,
  data: Uint8Array,
): Promise<void> {
  if (target === CLOSED) return
  const scope = toScope(target)
  let existing: Uint8Array = new Uint8Array(0)
  try {
    const [prior] = await dispatch('read', scope)
    existing = await materialize(prior as ByteSource)
  } catch (err) {
    if (!isFsError(err)) throw err
  }
  const combined: Uint8Array<ArrayBuffer> = new Uint8Array(existing.byteLength + data.byteLength)
  combined.set(existing, 0)
  combined.set(data, existing.byteLength)
  try {
    await dispatch('write', scope, [combined])
    session.execOpened.add(target)
  } catch (err) {
    if (!isFsError(err)) throw err
  }
}

/**
 * The `exec` arm. The redirect-only form is intercepted where redirects
 * are applied; a bare `exec` here has none, and `exec cmd` is the
 * process-replacement form this refuses.
 */
export function execBuiltin(call: BuiltinCall): Promise<Result> {
  return Promise.resolve(handleExecCommand([...call.argv.args], call.session))
}

File: python/mirage/workspace/executor/builtins/exec/exec.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import logging

from mirage.io import IOResult
from mirage.io.stream import materialize
from mirage.io.types import ByteSource
from mirage.runtime.types import DispatchFn
from mirage.shell.constants import (FD_BOTH, FD_CLOSE, FD_STDERR, FD_STDIN,
                                    FD_STDOUT)
from mirage.shell.descriptors import (bad_descriptor_line,
                                      unsupported_descriptor)
from mirage.shell.types import Redirect, RedirectKind
from mirage.types import PathSpec
from mirage.utils.errors import FS_ERRORS, fs_strerror
from mirage.workspace.executor.builtins.exec.constants import CLOSED
from mirage.workspace.executor.builtins.scope import _to_scope
from mirage.workspace.executor.builtins.types import BuiltinCall, Result
from mirage.workspace.executor.create import create_file
from mirage.workspace.session import Session
from mirage.workspace.types import ExecutionNode

logger = logging.getLogger(__name__)


async def handle_exec_command(
    args: list[str],
    session: Session,
) -> tuple[ByteSource | None, IOResult, ExecutionNode]:
    """The `exec` builtin without redirects.

    Bare `exec` is a no-op that succeeds. `exec CMD ...` asks the shell
    to replace itself with a program, which has no referent here (the
    in-process shell is an async executor, not an OS process: no PID,
    no `execve`), so it is refused loudly rather than run-then-exit,
    which would look like success while meaning something else. The
    redirect-only form (`exec > file`) never reaches here: it is a
    redirected statement, handled where redirects are applied.

    Args:
        args (list[str]): the words after `exec`.
        session (Session): shell session state.
    """
    if not args:
        return None, IOResult(), ExecutionNode(command="exec", exit_code=0)
    err = (f"mirage: exec: {args[0]}: process replacement is not supported "
           "(no OS process to replace)\n").encode()
    return None, IOResult(exit_code=2,
                          stderr=err), ExecutionNode(command="exec",
                                                     exit_code=2,
                                                     stderr=err)


async def install_exec_redirects(
    dispatch: DispatchFn,
    session: Session,
    redirects: list[Redirect],
) -> tuple[ByteSource | None, IOResult, ExecutionNode]:
    """Point the shell's own streams at files for the rest of the shell.

    The redirect-only `exec` form: `exec > file` sends every later
    statement's stdout to `file`, `exec 2> file` its stderr, `exec <
    file` feeds its stdin, and `exec >> file` appends. `2>&1` and `>&2`
    copy one stream's current target onto the other, and `>&-` / `<&-`
    close one. The output file is opened (created, and truncated unless
    appending) now, as bash opens it at `exec` time, so `exec > f`
    leaves an empty `f` even if nothing is written afterwards. A target
    that cannot be opened is bash's shell-attributed error and leaves
    the redirects unchanged. So is a descriptor above 2 (`exec 3>f`,
    `exec 3>&-`): the shell has no descriptor table, so the line is
    refused with `3: Bad file descriptor` rather than aliased onto
    stdout, which is what `exec 3>&-` used to close.

    Args:
        dispatch (DispatchFn): op dispatcher.
        session (Session): shell session state.
        redirects (list[Redirect]): the expanded redirects.
    """
    bad_fd = unsupported_descriptor(redirects)
    if bad_fd is not None:
        return _exec_failure(bad_descriptor_line(bad_fd))
    for r in redirects:
        if isinstance(r.target, int):
            # Keyed on the descriptor claimed, not the operator's
            # direction: `2<&-` closes stderr and `0>&-` stdin, as in
            # bash. `<&0` restores stdin; a dup of a descriptor onto
            # itself changes nothing.
            if r.fd == FD_STDIN:
                session.exec_stdin = b"" if r.target == FD_CLOSE else None
            elif r.target == FD_CLOSE:
                if r.fd == FD_STDERR:
                    session.exec_stderr = CLOSED
                else:
                    session.exec_stdout = CLOSED
            elif r.fd == FD_STDERR and r.target == FD_STDOUT:
                session.exec_stderr = session.exec_stdout
                session.exec_stderr_append = session.exec_stdout_append
            elif r.fd == FD_STDOUT and r.target == FD_STDERR:
                session.exec_stdout = session.exec_stderr
                session.exec_stdout_append = session.exec_stderr_append
            continue
        if ((r.kind == RedirectKind.STDIN) != (r.fd == FD_STDIN)):
            return _exec_failure(bad_descriptor_line(r.fd))
        scope = _to_scope(r.target) if isinstance(r.target, str) else r.target
        if r.kind == RedirectKind.STDIN:
            try:
                data, _ = await dispatch("read", scope)
            except FS_ERRORS as exc:
                return _exec_error(scope.raw_path, exc)
            session.exec_stdin = await materialize(data) or b""
            continue
        path = scope.virtual
        try:
            if await _open_target(dispatch, session, scope, r.append):
                session._exec_opened.add(path)
        except FS_ERRORS as exc:
            return _exec_error(scope.raw_path, exc)
        streams = ((["stderr"] if r.fd == FD_STDERR else ["stdout"])
                   if r.fd != FD_BOTH else ["stdout", "stderr"])
        for stream in streams:
            setattr(session, f"exec_{stream}", path)
            setattr(session, f"exec_{stream}_append", r.append)
    return None, IOResult(), ExecutionNode(command="exec", exit_code=0)


async def _open_target(dispatch: DispatchFn, session: Session, scope: PathSpec,
                       append: bool) -> bool:
    """Open an `exec` redirect target, the way bash does at `exec` time.

    Truncating creates the file empty; appending creates it only when it
    is not already there, so an existing one keeps its bytes. Either way
    the file exists before the next statement runs, which is what makes
    `exec >> new; test -e new` succeed with nothing written. Returns
    whether it was written, which is what marks the target opened.

    Args:
        dispatch (DispatchFn): op dispatcher.
        session (Session): the session holding the umask.
        scope (PathSpec): the target.
        append (bool): whether the redirect is `>>`.
    """
    if append:
        try:
            await dispatch("stat", scope)
            return False
        except FS_ERRORS as exc:
            logger.debug("exec append target %s is new: %s", scope.raw_path,
                         exc)
    await create_file(dispatch, session, scope, b"")
    return True


def _exec_error(label: str,
                exc: OSError) -> tuple[None, IOResult, ExecutionNode]:
    strerror = fs_strerror(exc)
    return _exec_failure(
        (f"{label}: {strerror}\n" if strerror else f"{label}\n").encode())


def _exec_failure(err: bytes) -> tuple[None, IOResult, ExecutionNode]:
    """The shell-attributed refusal of an `exec` redirect line.

    Args:
        err (bytes): the diagnostic, already in the shell's voice.
    """
    return None, IOResult(exit_code=1,
                          stderr=err), ExecutionNode(command="exec",
                                                     exit_code=1,
                                                     stderr=err)


async def divert_statement(
    dispatch: DispatchFn,
    session: Session,
    stdout: bytes | None,
    io: IOResult,
) -> bytes | None:
    """Send one statement's output to the shell's `exec` targets.

    Called after each top-level statement when an `exec` redirect is in
    force: stdout and stderr that point at a file are appended to it
    (the first write to each target having truncated it at `exec`
    time), and the stream is cleared so it does not also reach the
    terminal. A closed stream is dropped. Returns the stdout that
    should still bubble up, which is None once it has been diverted.

    Args:
        dispatch (DispatchFn): op dispatcher.
        session (Session): shell session state.
        stdout (bytes | None): the statement's materialized stdout.
        io (IOResult): the statement's result; its stderr is diverted
            in place.
    """
    if session.exec_stdout is not None and stdout:
        await _append(dispatch, session, session.exec_stdout, stdout)
        stdout = None
    elif session.exec_stdout is not None:
        stdout = None
    if session.exec_stderr is not None and io.stderr is not None:
        data = await materialize(io.stderr) or b""
        if data:
            await _append(dispatch, session, session.exec_stderr, data)
        io.stderr = None
    return stdout


async def _append(dispatch: DispatchFn, session: Session, target: str,
                  data: bytes) -> None:
    """Append bytes to an `exec` target, or drop them if it is closed.

    Args:
        dispatch (DispatchFn): op dispatcher.
        session (Session): shell session state.
        target (str): the target path, or `""` for a closed stream.
        data (bytes): the bytes to write.
    """
    if target == CLOSED:
        return
    scope = _to_scope(target)
    # The target exists by now, since `exec` opened it: every write is
    # read-then-append. A file deleted since then reads empty rather
    # than failing, which is where the debug line below comes from.
    existing = b""
    try:
        prior, _ = await dispatch("read", scope)
        existing = await materialize(prior) or b""
    except FS_ERRORS as exc:
        logger.debug("exec append pre-read failed for %s: %s", target, exc)
    try:
        await dispatch("write", scope, data=existing + data)
        session._exec_opened.add(target)
    except FS_ERRORS as exc:
        logger.debug("exec write failed for %s: %s", target, exc)


async def exec_builtin(call: BuiltinCall) -> Result:
    """The ``exec`` arm.

    The redirect-only form is intercepted where redirects are applied;
    a bare ``exec`` reaching here has no redirects, and ``exec cmd`` is
    the process-replacement form this refuses.

    Args:
        call (BuiltinCall): the invocation.
    """
    return await handle_exec_command(list(call.argv.args), call.session)
